### PR TITLE
refactor(core): migrate to standard Java ServiceLoader for plugins

### PR DESCRIPTION
**Pull request by fhussonnois**
Changes:
- add new top level interface `Plugin`
- add new `JsonDeserializer` aware of the `PluginRegistry`
- add new annotation processor for plugin
- migrate to hibernate-validator for the JSR-303 (no bean introspection dependencies)
- update all Validation annotations to be complitant with the JSR-303
- remove all KestraApplicationContext stuffs
- fix some tests

part-of: kestra-io/plugin-template-test#25

### CODE REVIEW

**Comment by fhussonnois** [comment]
**Issues**
* Issue on the `Pause` due to the `@NotEmpty` constraint on the inherited class `Sequential`

**Regression**
* Autocompletion for types in UI due to JSon Schema changes

**Limitations**
* Micronaut `@Controller` are not anymore considered as plugins

File: core/src/main/java/io/kestra/core/models/conditions/types/ExecutionFlowCondition.java [hunk 1]
```diff
@@ -37,6 +38,7 @@
         )
     }
 )
+@JsonDeserialize
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/models/conditions/types/ExecutionFlowCondition.java]: Is it still needed as not all conditions have it?

File: core/src/main/java/io/kestra/core/docs/Plugin.java [hunk 2]
```diff
@@ -57,31 +58,29 @@ public static Plugin of(RegisteredPlugin registeredPlugin) {
...
-            .filter(s -> !s.getName().startsWith("org.kestra."));
-    }
-
-    @SuppressWarnings("rawtypes")
-    private static <T> List<String> className(Class[] classes) {
-        return Arrays.stream(classes)
+            .filter(not(io.kestra.core.plugins.Plugin::isInternal))
             .map(Class::getName)
-            .collect(Collectors.toList());
+            .filter(c -> !c.startsWith("org.kestra."))
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/docs/Plugin.java:83]: What is it for? Is it already existing somewhere else?

File: core/src/main/java/io/kestra/core/plugins/PluginScanner.java [hunk 3]
```diff
@@ -82,78 +82,54 @@ public RegisteredPlugin scan() {
...
-                    definition.getName(), externalPlugin  != null ? externalPlugin.getLocation().toString() : (manifest != null ? manifest.getMainAttributes().getValue("X-Kestra-Title") : ""),
-                    e
-                );
-                continue;
+        final ServiceLoader<Plugin> sl = ServiceLoader.load(Plugin.class, classLoader);
+        sl.forEach(plugin -> {
+            // TODO - Manage internal plugin
+            /**
+            if(pluginType.isAnnotationPresent(Hidden.class)) {
+                return;
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/plugins/PluginScanner.java]: Please keep this so we can remove class incorrectly detected as plugins

File: core/src/main/java/io/kestra/core/tasks/flows/Sequential.java [hunk 4]
```diff
@@ -69,8 +69,8 @@ public class Sequential extends Task implements FlowableTask<VoidOutput> {
     protected List<Task> errors;
 
     @Valid
-    @NotEmpty
     @PluginProperty
+    // FIXME -> issue with Pause @NotEmpty
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/tasks/flows/Sequential.java:73]: I don't understand, this is overriden in the Pause task so it should work?

File: core/src/test/java/io/kestra/core/docs/JsonSchemaGeneratorTest.java [hunk 5]
```diff
@@ -281,7 +276,8 @@ private static abstract class ParentClass extends Task {
         beta = true,
         examples = {}
     )
-    private static class BetaTask extends Task {
+    @JsonDeserialize
```
Comment (loicmathieu) [inline core/src/test/java/io/kestra/core/docs/JsonSchemaGeneratorTest.java]: Should not be needed

File: model/src/main/java/io/kestra/core/plugins/PluginRegistry.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
+package io.kestra.core.plugins;
+
+import java.nio.file.Path;
+
+/**
+ * Registry for managing all Kestra's {@link Plugin}.
+ */
+public interface PluginRegistry {
```
Comment (loicmathieu) [inline model/src/main/java/io/kestra/core/plugins/PluginRegistry.java]: I don't think it should be in the model library

File: model/src/main/java/io/kestra/core/plugins/serdes/PluginDeserializer.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Specific {@link JsonDeserializer} for deserializing classes that implements the {@link Plugin} interface.
+ * <p>
+ * The {@link PluginDeserializer} uses the {@link PluginRegistry} to found the plugin class corresponding to
+ * a plugin type.
+ */
+public final class PluginDeserializer<T extends Plugin> extends JsonDeserializer<T> {
```
Comment (loicmathieu) [inline model/src/main/java/io/kestra/core/plugins/serdes/PluginDeserializer.java]: I don't think it should be in the model library

File: model/src/main/java/io/kestra/core/plugins/serdes/PluginDeserializer.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+    @SuppressWarnings("unchecked")
+    private static <T extends Plugin> T fromObjectNode(JsonParser jp,
+                                                       JsonNode node,
+                                                       DeserializationContext context) throws IOException {
+        Class<? extends Plugin> pluginType = null;
+
+        final String identifier = extractPluginRawIdentifier(node);
+        if (identifier != null) {
+            log.debug("Looking for Plugin for: {}",
```
Comment (loicmathieu) [inline model/src/main/java/io/kestra/core/plugins/serdes/PluginDeserializer.java]:
```suggestion
            log.trace("Looking for Plugin for: {}",
```

File: core/src/main/java/io/kestra/core/validations/CronExpression.java [hunk 9]
```diff
@@ -2,10 +2,14 @@
...
+import jakarta.validation.Payload;
+
 import java.lang.annotation.*;
 
 @Retention(RetentionPolicy.RUNTIME)
 @Constraint(validatedBy = CronExpressionValidator.class)
 public @interface CronExpression {
     String message() default "invalid cron expression ({validatedValue})";
+    Class<?>[] groups() default {};
+    Class<? extends Payload>[] payload() default {};
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/validations/CronExpression.java:14]: Can you explain what's the purpose of this?

**Review by loicmathieu — CHANGES_REQUESTED** [review]
Overall LGTM, but there is still some small things that I think should be improved:

- I'm not sure the PluginRegistry needs an interface.
- Can you explain the changes to validation, as it's a breaking change that would impact existing plugins
- Can you explain the Gradle change with the micronaut scope
- I have the feeling that the core vs external plugins can be simplified if not removed, this could be done later

I'll try this PR soon, one last thing that is missing is performance comparison. Starting Kestra when there is a lot of plugin takes some time, it would be interested to check if this PR improve or not this area.

**Comment by loicmathieu** [comment]
One last thing, split packages must be avoided if possible, so when moving stuff on the model library we should move all the classes of the package in it if possible.

**Pull request merged**